# Análisis de datos con Pyhon - Pre-entrega
Nombre y apellido:

# Etapa 1

## Script básico
Actividades 2 y 3

In [1]:
# Generamos una lista de diccionarios que contenga algunos datos
ventas = [
  {'producto': 'notebook', 'precio': 1500.0, 'cantidad': 3},
  {'producto': 'monitor', 'precio': 500.0, 'cantidad': 5},
  {'producto': 'mouse', 'precio': 20.0, 'cantidad': 15}
]

In [2]:
# Calculamos las ventas acumuladas - map
venta_acumuladas = sum(map(lambda venta: venta["precio"] * venta["cantidad"], ventas))
venta_acumuladas

7300.0

In [3]:
# Calculamos las ventas acumuladas - listas por comprensión
venta_acumuladas = sum([venta["precio"] * venta["cantidad"] for venta in ventas])
venta_acumuladas

7300.0

## Import Datasets
Actividad 1

In [4]:
import pandas as pd

In [5]:
# Import dataset ventas
df_ventas = pd.read_csv("https://github.com/alex-degarate/TT-2C2025-Data-Analitycs-Notebooks/raw/refs/heads/main/PreEntrega/Datasets/ventas.csv")


In [6]:
# Import dataset clientes
df_clientes = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/clientes.csv")

In [7]:
# Import dataset marketing
df_marketing = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/marketing.csv")

In [8]:
df_ventas

id_venta                producto   precio  cantidad fecha_venta  \
0          792       Cuadro decorativo   $69.94       5.0  02/01/2024   
1          811         Lámpara de mesa  $105.10       5.0  02/01/2024   
2         1156                Secadora   $97.96       3.0  02/01/2024   
3         1372                Heladera  $114.35       8.0  02/01/2024   
4         1546                Secadora  $106.21       4.0  02/01/2024   
...        ...                     ...      ...       ...         ...   
3030      1837         Horno eléctrico  $104.12       9.0  30/12/2024   
3031      2276                  Laptop   $85.27       9.0  30/12/2024   
3032      2696                  Laptop  $107.81       4.0  30/12/2024   
3033      2913              Smartphone   $99.85       7.0  30/12/2024   
3034      2930  Consola de videojuegos   $55.47       6.0  30/12/2024   

              categoria  
0            Decoración  
1            Decoración  
2     Electrodomésticos  
3     Electrodomésticos  
4     Electrodomésticos  
...                 ...  
3030  Electrodomésticos  
3031        Electrónica  
3032        Electrónica  
3033        Electrónica  
3034        Electrónica  

[3035 rows x 6 columns]

## Dataframe ventas

### Análisis exploratorio de **datos**
Actividad 4

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [9]:
# Aplicamos info() para familiarizarnos con el dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [10]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_ventas.head()

id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos

In [11]:
# Usar sample() para tomar registros al azar
df_ventas.sample(5)

id_venta             producto  precio  cantidad fecha_venta  \
1323      1721   Freidora eléctrica  $26.06       5.0  26/05/2024   
1147      1304           Aspiradora  $99.07       8.0  06/05/2024   
3009      2951  Parlantes Bluetooth  $89.89       6.0  27/12/2024   
1879      1405             Lavadora  $68.06      10.0  05/08/2024   
1802      1561   Freidora eléctrica  $88.22      12.0  28/07/2024   

              categoria  
1323  Electrodomésticos  
1147  Electrodomésticos  
3009        Electrónica  
1879  Electrodomésticos  
1802  Electrodomésticos

In [12]:
# Analizamos el campo precio
# Si es numericos se puede aplicar describe()
df_ventas["precio"].sample(5)

2244     $67.94
2645    $100.12
2547     $58.70
470      $68.89
2409     $91.96
Name: precio, dtype: object

In [13]:
# Analizamos el campo cantidad
# si es numerico se puede aplicar describe()
df_ventas["cantidad"].sample(5)

1384     5.0
2747    10.0
600      5.0
2343     1.0
1383     6.0
Name: cantidad, dtype: float64

In [14]:
# Analizamos el campo fecha_venta
df_ventas["fecha_venta"].sample(5)

2878    11/12/2024
2174    12/09/2024
1905    09/08/2024
1324    26/05/2024
2797    30/11/2024
Name: fecha_venta, dtype: object

In [15]:
# Analizamos el campo categoria
# Probar con value_counts y unique
df_ventas["categoria"].value_counts()

categoria
Decoración           1015
Electrónica          1012
Electrodomésticos    1008
Name: count, dtype: int64

### Calidad de datos
Actividad 5

In [16]:
# Buscamos registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [18]:
# Visualizamos los duplicados
# keep: 'first', 'last', False
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="id_venta")

id_venta           producto  precio  cantidad fecha_venta    categoria
820        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
826        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
832       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
844       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
878       204  Jarrón decorativo  $94.75       1.0  08/04/2024   Decoración
..        ...                ...     ...       ...         ...          ...
830      2545        Auriculares  $32.81      11.0  05/04/2024  Electrónica
883      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
889      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
831      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica
825      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica

[70 rows x 6 columns]

In [19]:
# Buscamos datos nulos
df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

In [29]:

# Visualizar los nulos
# df_ventas[df_ventas.isnull().any(axis=1)]
nulos_lista = df_ventas[df_ventas.isnull().any(axis=1)]
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [38]:
# Hay 2 registros con nulos, en el mismo registro, los campos "precio" y "cantidad son nulos
# No se pueden rescatar y van a ser eliminados
nulos_numero  = len(nulos_lista)
nulos_porcentaje = nulos_numero * 100 / len(df_ventas)
print(f"Numero de nulos: {nulos_numero } registros")
print(f"Porcentaje de nulos: {nulos_porcentaje:.2f}% registros")


Numero de nulos: 2 registros
Porcentaje de nulos: 0.07% registros


In [39]:
# Analizar los precios de los NaN a ver que opciones tenemos
# df_ventas[df_ventas["producto"]=="Elementos de cerámica"]
df_ventas["producto"].value_counts()
# df_ventas["producto"].unique()


producto
Lámpara de mesa           181
Auriculares               147
Microondas                135
Cafetera                  118
Laptop                    104
Secadora                  103
Jarrón decorativo         102
Smartphone                102
Alfombra                  102
SmartWatch                102
Aspiradora                102
Cortinas                  102
Rincón de plantas         102
Freidora eléctrica        101
Espejo decorativo         101
Adorno de pared           101
Lavadora                  100
Consola de videojuegos    100
Cuadro decorativo         100
Horno eléctrico           100
Plancha de vapor          100
Proyector                 100
Elementos de cerámica     100
Batidora                  100
Televisor                 100
Cámara digital            100
Parlantes Bluetooth        99
Tablet                     58
Heladera                   49
Candelabro                 24
Name: count, dtype: int64

### Observaciones




* Cantidad de registros originales: 3035
* Columnas/campos originales: 6
* 35 registros duplicados &emsp; => eliminarlos
*  2 registros con datos nulos: id:[136, 139], < 0.1 % total &emsp; =>  dropna
* Campo precio: tipo objeto con `$`, removerlo y luego cambiar el dtype a float64
* Campo fecha: ajustar a dtype datetime

## Dataframe marketing

### Analisis exploratorio de datos

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [40]:
# Aplicamos info() para familiarizarnos con el dataframe
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


In [48]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
pd.set_option('display.max_rows', None)
df_marketing.head()

id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024

In [44]:
# Analizamos el campo costo
df_marketing["costo"].sample(5)

43    5.65
5     5.07
87    5.81
79    4.99
31    3.22
Name: costo, dtype: float64

In [45]:
# Analizamos el campo fecha_inicio
df_marketing["fecha_inicio"].sample(5)

85    13/12/2024
87    17/12/2024
14    09/05/2024
0     20/03/2024
28    20/06/2024
Name: fecha_inicio, dtype: object

In [46]:
# Analizamos el campo fecha_fin
df_marketing["fecha_fin"].sample(5)

62    02/11/2024
49    20/09/2024
31    26/07/2024
11    17/06/2024
2     20/04/2024
Name: fecha_fin, dtype: object

### Calidad de los datos


In [ ]:
# Buscamos registros duplicados


### Observaciones
El campo fecha_inicio es objeto &emsp; => cambiar a dtype datetime

El campo fecha_fin &emsp; es objeto &emsp; => cambiar a dtype datetime


## Dataframe clientes

### Analisis exploratorio de datos

### Observaciones

# Etapa 2

## Limpieza de datos
Actividad 1 - Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.

### Dataframe ventas

* 35 duplicados, eliminarlos - se eliminaron 35 duplicados &emsp; => Ok
* Algunos nulos (2), 1% del faltante, dropna - se eliminaron 2 registros con 2 NaN
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64 - Ok
* Campo fecha: ajustar a dtype datetime - Ok

In [49]:
# Eliminamos registros duplicados
# atributos a evaluar, keep, ignore_index, inplace
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [50]:
# Validamos que no haya duplicados
df_ventas.duplicated().sum()

np.int64(0)

In [51]:
# Eliminamos registros con datos nulos
df_ventas.dropna(inplace=True)

In [52]:
# Validamos que no haya nulos
# df_ventas.isnull().sum()
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2998 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     2998 non-null   int64  
 1   producto     2998 non-null   object 
 2   precio       2998 non-null   object 
 3   cantidad     2998 non-null   float64
 4   fecha_venta  2998 non-null   object 
 5   categoria    2998 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 164.0+ KB


In [53]:
# Eliminamos el caracter especial $ de precio
# Normalización: usar replace()
# De ser necesario, podemos usar el atributo regex=True
df_ventas["precio"] = df_ventas["precio"].str.strip().str.replace("$", "")

In [54]:
# Validamos que precio no contenga $
df_ventas["precio"].sample(3)

409      98.73
2489     91.06
387     101.08
Name: precio, dtype: object

In [55]:
# Convertimos precio de str a float
# Podemos usar pd.to_numeric o astype o incluso combinarlos
df_ventas["precio"] = df_ventas["precio"].astype(float)


In [56]:
# Convertir el dtype de cantidad, de float64 a int64
df_ventas["cantidad"]= df_ventas["cantidad"].astype(int)

In [69]:
# Validar la conversion a int
df_ventas["cantidad"].dtypes


dtype('int64')

In [68]:
# Validamos campos numericos
df_ventas[["precio","cantidad"]].sample(3)

precio  cantidad
2386  111.81         9
1755   57.13         8
208    39.45         9

In [ ]:
# Convertir el campo fecha_venta a dtype datetime
# pd.to_datetime(df_ventas["fecha_venta"], format="%d/%m/%Y")
# pd.to_datetime(df_ventas["fecha_venta"], dayfirst=True)
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [ ]:
# Validar la conversion de fecha
df_ventas['fecha_venta'].dtype

### Dataframe marketing

*** Acciones ***

In [ ]:
# Convertir el campo fecha_inicio a dtype datetime


In [ ]:
# Convertir el campo fecha_fin a dtype datetime


In [ ]:
# Validamos los cambios


## Transformación de datos
Actividad 2 - Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.

In [ ]:
df_ventas.head(2)

In [ ]:
# Transformación:
# Calculamos el valor de cada venta = precio * cantidad
# Agregamos una nueva columnas
df_ventas["valor_venta"] = df_ventas["precio"] * df_ventas["cantidad"]

In [ ]:
# Visualizamos
df_ventas.head(2)

In [ ]:
# Agregación:
# Agrupar valor_venta por producto
# Sumamos las ventas dentro de cada grupo
tabla_ventas = (df_ventas.groupby("producto", as_index=False)["valor_venta"].sum())
tabla_ventas.head()

In [ ]:
# Usemos describe para analizar algunos valores del dataframe resultante
tabla_ventas.describe()

In [ ]:
# Agregamos un plot para visualizar graficamente la distribución
import seaborn as sns
import matplotlib.pyplot as plt

sns.histplot(tabla_ventas["valor_venta"], bins=30, kde=True, edgecolor="black")
plt.title("Distribución de Valor de Venta con Curva de Densidad")
plt.xlabel("Valor de venta")
plt.ylabel("Frecuencia")
plt.show()

In [ ]:
# Calcular el percentil 75 (umbral de alto rendimiento)
umbral_ventas = tabla_ventas["valor_venta"].quantile(0.75)
umbral_ventas

In [ ]:
# Identificar productos que superan el umbral
# productos_top = tabla_ventas.query("valor_venta >= @umbral")
productos_top = tabla_ventas[tabla_ventas["valor_venta"] >= umbral_ventas]
productos_top

## Agregación
Actividad 3 - Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [ ]:
df_ventas.columns

In [ ]:
# Agrupamos por categoria y agregamos la suma
ventas_categoria = df_ventas.groupby("categoria", as_index=False)["valor_venta"].sum()
ventas_categoria.head()

In [ ]:
ventas_categoria

In [ ]:
# Ordenar y mostrar los resultados
ventas_categoria

## Integración de datos
Actividad 4 - Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

In [ ]:
df_ventas.head(2)

In [ ]:
df_marketing.info()

In [ ]:
# Unir los datasets ventas y marketing por columna 'producto'
ventas_marketing = pd.merge(df_ventas, df_marketing, on="producto", how="left")

In [ ]:
# Verificar coincidencias (muestra)
ventas_marketing.head()

### Analizamos el impacto de la campaña

In [ ]:
# Nos quedamos solo con los registros de ventas durante la campaña de marketing
# Seleccionamos solo los campos que nos interesan: "producto", "valor_venta", "costo"


In [ ]:
# Agrupamos por producto y hacemos una agregación por "valor de venta" y "costo"
ventas_marketing.groupby("categoria")[["valor_venta","costo"]].sum()

In [ ]:
# Generamos una nueva columna con la diferencia entre lo vendido y lo invertido en la campaña
